_posts/2021-09-23-event-streaming-with-pulsar-and-scala.md

```diff
@@ -0,0 +1,147 @@
+5. **Tiered Storage** As you process enormous volumes of data your topic might grow large without limit, which can become quite expensive over time. Apache Pulsar offers tiered storage and so as your topics grow, you can offload older data to some cheaper storage like Amazon S3, while your clients can still access the data and continue serving it as if nothing had changed. 
+6. **Pulsar Functions** Pulsar Functions are a lightweight serverless compute framework that allow you to deploy your own stream processing logic in a very simple way. Being lightweight it also makes it an excellent choice for iot edge analytics use cases. Combined with Function Mesh (https://functionmesh.io/) a new project recently open sourced by Streamnative (https://streamnative.io/) is what makes event stream processing truly cloud native by deploying everything as part of a mesh and leveraging native kubernetes features like autoscaling. 
+
+There are more features of Apache Pulsar like a built-in Schema Registry, support for transactions and Pulsar SQL, but at this point lets see how you can actually get Pulsar up and running and create our very first Producers and Consumers in Scala.
+
+## 2. Example Use Case and Cluster Setup
+As a simple example use case we will create a Producer that simulates sensor event readings, sends them over to a topic and then on the other end create a consumer that subscribes to that topic and just reads the incoming events. We will be using the [pulsar4s](https://github.com/sksamuel/pulsar4s) scala library for our implementation and we will run a Pulsar cluster using docker. In order to start a Pulsar cluster in standalone mode, run the following command within your terminal:
+```shell
+docker run -it \
+  -p 6650:6650 \
+  -p 8080:8080 \
+  --name pulsar \
+  apachepulsar/pulsar:2.8.0 \
+  bin/pulsar standalone
+```
+
+This command will start pulsar and bind the necessary ports to your local machine. With our cluster up and running, let's start creating our producers and consumers.
+
+## 3. Apache Pulsar Producers
+First, we need the dependencies for the pulsar4s-core and pulsar4s-circe:
+```scala
+val pulsar4sVersion = "2.7.3"
+
+lazy val pulsar4s       = "com.sksamuel.pulsar4s" %% "pulsar4s-core"  % pulsar4sVersion
+lazy val pulsar4sCirce  = "com.sksamuel.pulsar4s" %% "pulsar4s-circe" % pulsar4sVersion
+
+libraryDependencies ++= Seq(
+  pulsar4s, pulsar4sCirce
+)
+```
+
+The main entry point for all producing and consuming applications is the Pulsar Client, which handles the connection to the brokers. From within the Pulsar client you can also set up the authentication to your cluster as well as other important tuning configurations such as timeout settings and connection pools. You can simply instantiate the client by providing the service url you want to connect to.
+
+```scala
+val pulsarClient = PulsarClient("pulsar://localhost:6650")
```

Comment:
add the appropriate imports with each new non-standard type, or add _all_ of them at the beginning